src/gammalasso.jl

```diff
@@ -0,0 +1,406 @@
+                            end
+                        end
+                    end
+                end
+
+                converged && break
+            end
+            converged || error("IRLS failed to converge in $irls_maxiter iterations at λ = $(curλ)")
+
+            dev_ratio = dev/nulldev
+
+            pct_dev[i] = 1 - dev_ratio
+            addcoefs!(coefs, newcoef, i)
+            b0s[i] = b0
+
+            # Test whether we should continue
+            if i == nλ || (autoλ && (last_dev_ratio - dev_ratio < MIN_DEV_FRAC_DIFF ||
+                           pct_dev[i] > MAX_DEV_FRAC))
+                break
+            end
+
+            verbose && println("$i: λ=$curλ, pct_dev=$(pct_dev[i])")
+            i += 1
+            computeω!(ω,γ,newcoef) # use β^{i-1} for β^{i} gamma lasso weights
+        end
+    end
+
+    path.λ = path.λ[1:i]
+    path.pct_dev = pct_dev[1:i]
+    path.coefs = coefs[:, 1:i]
+    path.b0 = b0s[1:i]
+    path.niter = niter
+    if !isempty(Xnorm)
+        scale!(Xnorm, path.coefs)
+    end
+end
+
+# Fits linear models (just the outer loop)
+function StatsBase.fit!{S<:LinearModel,T}(path::GammaLassoPath{S,T}; verbose::Bool=false,
+                                          cd_maxiter::Int=10000, cd_tol::Real=1e-7, irls_tol::Real=1e-7,
+                                          criterion=:coef, minStepFac::Real=eps())
+    0 < minStepFac < 1 || error("minStepFac must be in (0, 1)")
+    criterion == :obj || criterion == :coef || error("criterion must be obj or coef")
+
+    @extractfields path nulldev λ autoλ Xnorm m γ
+    nλ = length(λ)
+    r = m.rr
+    cd = m.pp
+    cd.maxiter = cd_maxiter
+    cd.tol = cd_tol
+
+    X = cd.X
+    coefs = spzeros(T, size(X, 2), nλ)
+    b0s = zeros(T, nλ)
+    newcoef = SparseCoefficients{T}(size(X, 2))
+    pct_dev = zeros(nλ)
+    dev_ratio = convert(T, NaN)
+    niter = 0
+
+    update!(cd, newcoef, r.mu, r.wts)
+
+    if autoλ
+        # No need to fit the first model
+        b0s[1] = path.nullb0
+        i = 2
+    else
+        i = 1
+    end
+
+    while true # outer loop
+        last_dev_ratio = dev_ratio
+        curλ = λ[i]
+
+        # Run coordinate descent
+        niter += cdfit!(newcoef, cd, curλ, criterion)
+
+        dev_ratio = cd.dev/nulldev
+        pct_dev[i] = 1 - dev_ratio
+        addcoefs!(coefs, newcoef, i)
+        b0s[i] = intercept(newcoef, cd)
+
+        # Test whether we should continue
+        if i == nλ || (autoλ && last_dev_ratio - dev_ratio < MIN_DEV_FRAC_DIFF ||
+                       pct_dev[i] > MAX_DEV_FRAC)
+            break
+        end
+
+        verbose && println("$i: λ=$curλ, pct_dev=$(pct_dev[i])")
+        i += 1
+        computeω!(cd.ω,γ,newcoef) # use β^{i-1} for β^{i} gamma lasso weights
+    end
+
+    path.λ = path.λ[1:i]
+    path.pct_dev = pct_dev[1:i]
+    path.coefs = coefs[:, 1:i]
+    path.b0 = b0s[1:i]
+    path.niter = niter
+    if !isempty(Xnorm)
+        scale!(Xnorm, path.coefs)
+    end
+end
```

Comment:
Similarly, since these are very similar to `fit` for LassoPath it would be good to avoid repeating all this code.